lib/internal/Magento/Framework/Lock/Backend/Database.php

```diff
@@ -84,12 +84,7 @@ public function lock(string $name, int $timeout = -1): bool
          * currently we support MySQL 5.6 way only.
          */
         if ($this->currentLock) {
-            throw new AlreadyExistsException(
-                new Phrase(
-                    'Current connection is already holding lock for %1, only single lock allowed',
-                    [$this->currentLock]
-                )
-            );
+            return false;
```

Comment:
This creates infinite loop if Database is used for AbstractBlock locking. All next `$this->locker->lock()` calls return false.

If a cacheable block has a child cacheable block, the `LockGuardedCacheLoader::lockedLoadData()` call is nested, but since there is already a lock in the Database backend, and it does not depend on the lock name, lock() returns false. See 2 `lockedLoadData()` in this call stack.
```
LockGuardedCacheLoader.php:111, Magento\Framework\Cache\LockGuardedCacheLoader->lockedLoadData()
AbstractBlock.php:1125, Amasty\Label\Block\Label->_loadCache()
AbstractBlock.php:674, Amasty\Label\Block\Label->toHtml()
...
Interceptor.php:429, Magento\Catalog\Block\Product\Image\Interceptor->toHtml()
grid.phtml:52, include()
...
Template.php:301, Magento\CatalogWidget\Block\Product\ProductsList\Interceptor->_toHtml()
AbstractBlock.php:1100, Magento\CatalogWidget\Block\Product\ProductsList\Interceptor->Magento\Framework\View\Element\{closure:/var/www/html/vendor/magento/framework/View/Element/AbstractBlock.php:1094-1101}()
LockGuardedCacheLoader.php:103, Magento\Framework\Cache\LockGuardedCacheLoader->lockedLoadData()
AbstractBlock.php:1125, Magento\CatalogWidget\Block\Product\ProductsList\Interceptor->_loadCache()
AbstractBlock.php:674, Magento\CatalogWidget\Block\Product\ProductsList\Interceptor->toHtml()
Interceptor.php:830, Magento\CatalogWidget\Block\Product\ProductsList\Interceptor->toHtml()
Filter.php:121, Magento\Widget\Model\Template\Filter\Interceptor->generateWidget()
Interceptor.php:24, Magento\Widget\Model\Template\Filter\Interceptor->generateWidget()
Filter.php:132, Magento\Widget\Model\Template\Filter\Interceptor->widgetDirective()
Interceptor.php:37, Magento\Widget\Model\Template\Filter\Interceptor->widgetDirective()
LegacyDirective.php:42, ReflectionMethod->invokeArgs()
LegacyDirective.php:42, Magento\Framework\Filter\DirectiveProcessor\LegacyDirective->process()
Template.php:184, Magento\Widget\Model\Template\Filter\Interceptor->filter()
Filter.php:1080, Magento\Widget\Model\Template\Filter\Interceptor->filter()
Interceptor.php:58, Magento\Widget\Model\Template\Filter\Interceptor->___callParent()
Interceptor.php:138, Magento\Widget\Model\Template\Filter\Interceptor->Magento\Framework\Interception\{closure:/var/www/html/vendor/magento/framework/Interception/Interceptor.php:104-151}()
Interceptor.php:153, Magento\Widget\Model\Template\Filter\Interceptor->___callPlugins()
Interceptor.php:403, Magento\Widget\Model\Template\Filter\Interceptor->filter()
Page.php:170, Magento\VersionsCms\Block\Cms\Page->_toHtml()
AbstractBlock.php:1100, Magento\VersionsCms\Block\Cms\Page->Magento\Framework\View\Element\{closure:/var/www/html/vendor/magento/framework/View/Element/AbstractBlock.php:1094-1101}()
AbstractBlock.php:1104, Magento\VersionsCms\Block\Cms\Page->_loadCache()
AbstractBlock.php:674, Magento\VersionsCms\Block\Cms\Page->toHtml()
...
Http.php:120, Magento\Framework\App\Http\Interceptor->launch()
Interceptor.php:24, Magento\Framework\App\Http\Interceptor->launch()
Bootstrap.php:261, Magento\Framework\App\Bootstrap->run()
index.php:40, {main}()
```

Reply:
Could you please report a separate issue with all details for verifying and  than fixing that?